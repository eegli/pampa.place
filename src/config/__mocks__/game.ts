import {GameConfig} from '../types';

export const config: GameConfig = {
  maxPlayers: 4,
  rounds: [1, 2, 3],
  roundsDefault: 2,
  timeLimits: [30, 60, 120, -1],
  timeLimitsDefault: 30,
};
