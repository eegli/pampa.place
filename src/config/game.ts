type GameConfig = {
  maxPlayers: number;
  roundSelect: [number, number, number];
  timeLimit: number;
};

const gameConfig: GameConfig = {
  maxPlayers: 4,
  roundSelect: [1, 3, 5],
  timeLimit: 10,
};

export default gameConfig;
