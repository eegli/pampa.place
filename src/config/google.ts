type GoogleConfig = {
  svRequest: google.maps.StreetViewLocationRequest;
  streetview: google.maps.StreetViewPanoramaOptions;
  map: google.maps.MapOptions;
};

export const config: GoogleConfig = {
  svRequest: {
    radius: 1000,
  },
  streetview: {
    zoom: 0,
    addressControl: false,
    motionTracking: false,
    zoomControl: false,
    showRoadLabels: false,
    fullscreenControl: false,
    pov: {
      heading: 35,
      pitch: 10,
    },
  },
  map: {
    draggableCursor: 'crosshair',
    clickableIcons: false,
    disableDefaultUI: true,
    mapTypeId: 'hybrid',
    mapTypeControl: true,
    gestureHandling: 'auto',
    mapTypeControlOptions: {
      mapTypeIds: ['hybrid', 'roadmap'],
    },
  },
};
