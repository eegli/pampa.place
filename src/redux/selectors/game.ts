import { createSelector } from '@reduxjs/toolkit';
import { RootState } from '../store';

export const getActivePlayer = (s: RootState) => s.game.players[0];

export const getPlayerCount = (s: RootState) => s.game.players.length;

// TODO use memoization once redux toolkit exports reselect 4.1
export const getRoundScores = createSelector(
  [
    (s: RootState) => s.game.scores,
    (s: RootState) => s.game.rounds.current - 1,
  ],
  (scores, currentRound) => {
    /* Sort on a copy, otherwise, Redux isn't happy: https://stackoverflow.com/a/66721870 */
    return [...scores[currentRound]].sort((a, b) =>
      a.score > b.score ? -1 : b.score > a.score ? 1 : 0
    );
  }
);

export const getTotalScores = createSelector(
  [(s: RootState) => s.game.scores, (s: RootState) => s.game.players],
  (scores, players) => {
    const final = Array.from({ length: players.length }, (_, idx) => ({
      name: players[idx],
      score: 0,
    }));

    scores.forEach(round => {
      round.forEach((playerResult, playerIdx) => {
        final[playerIdx].score += playerResult.score;
      });
    });

    return final.sort((a, b) =>
      a.score > b.score ? -1 : b.score > a.score ? 1 : 0
    );
  }
);

const s = [
  [
    { name: 'e', round: 1, selected: 'x', dist: 1, score: 999 },
    { name: 'x', round: 1, selected: 'x', dist: 1, score: 999 },
  ],
  [
    { name: 'e', round: 2, selected: 'x', dist: 1, score: 999 },
    { name: 'x', round: 2, selected: 'x', dist: 1, score: 999 },
  ],
];

export const shouldRequestNewSV = (s: RootState) =>
  s.game.rounds.progress === 0;

export const isFinished = (s: RootState) =>
  s.game.rounds.progress === s.game.players.length &&
  s.game.rounds.current === s.game.rounds.total;
