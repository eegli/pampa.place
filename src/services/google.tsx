import {unsafeToggleHTMLElement} from '@/utils/misc';

export const GstvContainer = () => {
  return (
    <div id="__GSTV__CONTAINER__" data-testid="__GSTV__CONTAINER__">
      <div id="__GSTV__" data-testid="__GSTV__" style={{height: '100%'}} />
    </div>
  );
};

export const GmapContainer = () => {
  return (
    <div id="__GMAP__CONTAINER__" data-testid="__GMAP__CONTAINER__">
      <div id="__GMAP__" data-testid="__GMAP__" style={{height: '100%'}} />
    </div>
  );
};

export class StreetViewService {
  static get div() {
    return document.getElementById('__GSTV__')!;
  }
  static get container() {
    return document.getElementById('__GSTV__CONTAINER__')!;
  }
  private static _sv: google.maps.StreetViewPanorama | undefined;

  static get sv() {
    if (!StreetViewService._sv) {
      StreetViewService._sv = new google.maps.StreetViewPanorama(
        StreetViewService.div
      );
      console.log('%cCreated new global Street View instance', 'color: green');
    }
    return StreetViewService._sv;
  }

  static mount<T extends HTMLElement>(tempContainer: T) {
    const unmount = unsafeToggleHTMLElement(
      StreetViewService.div,
      StreetViewService.container,
      tempContainer
    );
    return () => {
      unmount();
    };
  }
}

export class MapService {
  static _map: google.maps.Map | undefined;
  static markers: google.maps.Marker[] = [];

  static addMarker(m: google.maps.Marker) {
    MapService.markers.push(m);
    return m;
  }
  // https://developers.google.com/maps/documentation/javascript/markers#remove
  static clearMarkers() {
    MapService.markers.forEach(m => m.setMap(null));
    MapService.markers.length = 0;
  }

  static get div() {
    return document.getElementById('__GMAP__')!;
  }
  static get container() {
    return document.getElementById('__GMAP__CONTAINER__')!;
  }

  static get map() {
    if (!MapService._map) {
      MapService._map = new google.maps.Map(MapService.div);
      console.log('%cCreated new global Map instance', 'color: green');
    }
    return MapService._map;
  }

  static mount<T extends HTMLElement>(tempContainer: T) {
    const unmount = unsafeToggleHTMLElement(
      MapService.div,
      MapService.container,
      tempContainer
    );

    return () => {
      unmount();
    };
  }
}
