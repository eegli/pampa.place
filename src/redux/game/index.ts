import {config} from '@/config/game';
import {MAPS, MAP_IDS} from '@/config/maps';
import {LatLngLiteral} from '@/config/types';
import {calcDist, calcScore} from '@/utils/geo';
import {OrNull} from '@/utils/types';
import {createSlice, PayloadAction} from '@reduxjs/toolkit';

export type Result = {
  name: string;
  selected: OrNull<LatLngLiteral>;
  dist: number;
  score: number;
};

export enum STATUS {
  UNINITIALIZED = 'UNINITIALIZED',
  ROUND_STARTED = 'ROUND_STARTED',
  PENDING_PLAYER = 'PENDING_PLAYER',
  ROUND_ENDED = 'ROUND_ENDED',
  FINISHED = 'FINISHED',
}

export interface GameState {
  status: STATUS;
  mapId: string;
  mapName: string;
  players: string[];
  // Store the results for each round in the outer array. The innner
  // array consists of player result objects.
  scores: Result[][];
  timeLimit: number;
  rounds: {
    current: number;
    progress: number;
    total: number;
  };
}

const initialState: GameState = {
  status: STATUS.UNINITIALIZED,
  mapId: MAP_IDS[0].id,
  mapName: MAP_IDS[0].name,
  players: [],
  scores: [],
  timeLimit: config.timeLimitsDefault,
  rounds: {
    current: 1,
    total: config.roundsDefault,
    progress: 0,
  },
};

const gameSlice = createSlice({
  name: 'game',
  initialState,

  // The active player is always at index 0, and players are rotated
  // when they are done
  reducers: {
    reset(state) {
      state = initialState;
    },
    setPlayers(state, action: PayloadAction<string[]>) {
      state.players = action.payload.filter(Boolean).map(el => el.slice(0, 24));
    },
    setRounds(state, action: PayloadAction<number>) {
      state.rounds.total = action.payload;
    },
    setTimeLimit(state, action: PayloadAction<number>) {
      state.timeLimit = action.payload;
    },
    setMap(state, action: PayloadAction<string>) {
      state.mapId = action.payload;
      state.mapName = MAPS[action.payload].feature.properties.name;
    },
    initGame(state) {
      if (!state.players.length) {
        state.players = ['Player 1'];
      }
      state.scores = [];
      // Empty score array for first round
      state.scores.push([]);
      state.rounds.current = 1;
      state.rounds.progress = 0;
      state.status = STATUS.PENDING_PLAYER;
    },
    startRound(state) {
      state.status = STATUS.ROUND_STARTED;
    },
    setPlayerScore(
      state,
      {
        payload,
      }: PayloadAction<{
        selected: OrNull<LatLngLiteral>;
        initial: OrNull<LatLngLiteral>;
      }>
    ) {
      // Payload when the user fails to set a location in time
      let score = 0;
      let dist = -1;
      let selected: OrNull<LatLngLiteral> = null;

      // User managed to set a location
      if (payload.selected && payload.initial) {
        dist = calcDist(payload.initial, payload.selected);
        score = calcScore(MAPS[state.mapId].area, dist);
        selected = payload.selected;
      }

      state.scores[state.rounds.current - 1].push({
        name: state.players[0],
        selected,
        dist,
        score,
      });

      // Set round progress
      state.rounds.progress++;

      // Rotate players
      const last = state.players.pop();
      if (last) state.players.unshift(last);

      // Same round, wait for player change
      if (state.rounds.progress < state.players.length) {
        // Display popup
        state.status = STATUS.PENDING_PLAYER;
        // Round is over, reset round progress
      } else {
        // Update overall score and force a new random location
        state.status = STATUS.ROUND_ENDED;
      }
    },
    resetRound(state) {
      // Reset score for this round
      state.scores[state.rounds.current - 1] = [];

      // Reset round progress
      state.rounds.progress = 0;

      // Rotate to first player again
      for (let i = 0; i < state.rounds.progress; i++) {
        const playerName = state.players.pop();
        if (playerName) {
          state.players.unshift(playerName);
        }
      }

      // Set zero for current round scores
      state.status = STATUS.PENDING_PLAYER;
    },
    endRound(state) {
      // Game has ended
      if (state.rounds.total === state.rounds.current) {
        state.status = STATUS.FINISHED;
      } else {
        // Prepare for next round
        state.rounds.current++;
        state.rounds.progress = 0;
        state.scores.push([]);
        state.status = STATUS.PENDING_PLAYER;
      }
    },
  },
});

export const {
  reset,
  initGame,
  startRound,
  endRound,
  resetRound,
  setPlayerScore,
  setRounds,
  setMap,
  setPlayers,
  setTimeLimit,
} = gameSlice.actions;
export default gameSlice;
