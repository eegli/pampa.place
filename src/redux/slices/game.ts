import { config } from '@/config/game';
import { CUSTOM_MAP_IDS, LatLngLiteral, MAPS } from '@/config/maps';
import { calcDist, calcScore } from '@/utils/geo';
import { OrNull } from '@/utils/types';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export type Result = {
  round: number;
  selected: OrNull<LatLngLiteral>;
  dist: number;
  score: number;
};

type Player = {
  totalScore: number;
} & { results: Result[] };

export enum STATUS {
  PENDING_START = 'PENDING_START', // Uninitialized
  ROUND_STARTED = 'ROUND_STARTED',
  INTERMISSION = 'INTERMISSION',
  ROUND_ENDED = 'ROUND_ENDED',
  FINISHED = 'FINISHED',
}

export interface GameState {
  status: STATUS;
  mapId: string;
  mapName: string;
  players: {
    names: string[];
    scores: Record<string, Player>;
  };
  timeLimit: number;
  rounds: {
    current: number;
    progress: number;
    total: number;
  };
}

export const initialState: GameState = {
  status: STATUS.PENDING_START,
  mapId: CUSTOM_MAP_IDS[0].id,
  mapName: CUSTOM_MAP_IDS[0].name,
  players: {
    names: [],
    scores: {},
  },
  timeLimit: config.timeLimitsDefault,
  rounds: {
    current: 1,
    total: config.roundsDefault,
    progress: 0,
  },
};

export const gameSlice = createSlice({
  name: 'game',
  initialState,

  // The active player is always at index 0, and players are rotated
  // when they are done
  reducers: {
    reset(state) {
      state = initialState;
    },
    setPlayers(state, action: PayloadAction<string[]>) {
      state.players.names = action.payload
        .filter(Boolean)
        .map(el => el.slice(0, 24));
    },
    setRounds(state, action: PayloadAction<number>) {
      state.rounds.total = action.payload;
    },
    setTimeLimit(state, action: PayloadAction<number>) {
      state.timeLimit = action.payload;
    },
    setMap(state, action: PayloadAction<string>) {
      state.mapId = action.payload;
      state.mapName = MAPS[action.payload].geo.properties.name;
    },
    initGame(state) {
      if (!state.players.names.length) {
        state.players.names = ['Player 1'];
      }

      state.players.scores = {};

      state.players.names.forEach(name => {
        state.players.scores[name] = { totalScore: 0, results: [] };
      });

      state.rounds.current = 1;
      state.rounds.progress = 0;

      state.status = STATUS.INTERMISSION;
    },
    finishRound(state) {
      state.rounds.current++;
      state.rounds.progress = 0;

      // Game has ended
      if (state.rounds.total < state.rounds.current) {
        state.status = STATUS.FINISHED;
      } else {
        state.status = STATUS.INTERMISSION;
      }
    },
    startRound(state) {
      state.status = STATUS.ROUND_STARTED;
    },
    resetRound(state) {
      // Rotate to first player again
      for (let i = 0; i < state.rounds.progress; i++) {
        const playerName = state.players.names.pop();
        if (playerName) {
          // Remove score from this round
          const result = state.players.scores[playerName].results.pop();

          // Subtract score from this round from total score
          if (result)
            state.players.scores[playerName].totalScore -= result.score;

          // Rotate player
          state.players.names.unshift(playerName);
        }
      }

      // Reset round progress
      state.rounds.progress = 0;

      // Set zero for current round scores
      state.status = STATUS.INTERMISSION;
    },
    setPlayerScore(
      state,
      {
        payload,
      }: PayloadAction<{
        selected: OrNull<LatLngLiteral>;
        initial: OrNull<LatLngLiteral>;
      }>
    ) {
      // Set current round score
      const player = state.players.scores[state.players.names[0]];

      // Payload when the user fails to set a location in time
      let score = 0;
      let dist = -1;
      let selected: OrNull<LatLngLiteral> = null;

      // User managed to set a location
      if (payload.selected && payload.initial) {
        dist = calcDist(payload.initial, payload.selected);
        score = calcScore(MAPS[state.mapId].computed.area, dist);
        selected = payload.selected;
      }

      const newScore = player.totalScore + score;
      const existingResults = player.results;
      state.players.scores[state.players.names[0]] = {
        totalScore: newScore,
        results: [
          ...existingResults,
          {
            round: state.rounds.current,
            selected,
            dist,
            score,
          },
        ],
      };

      // Set round progress
      state.rounds.progress++;

      // Rotate players
      const last = state.players.names.pop();
      if (last) state.players.names.unshift(last);

      // Same round, wait for player change
      if (state.rounds.progress < state.players.names.length) {
        // Display popup
        state.status = STATUS.INTERMISSION;
        // Round is over, reset round progress
      } else {
        // Update overall score and force a new random location
        state.status = STATUS.ROUND_ENDED;
      }
    },
  },
});

export const {
  reset,
  initGame,
  startRound,
  finishRound,
  resetRound,
  setPlayerScore,
  setRounds,
  setMap,
  setPlayers,
  setTimeLimit,
} = gameSlice.actions;
export default gameSlice.reducer;
