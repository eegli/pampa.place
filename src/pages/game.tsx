import { NextPage } from 'next';
import { useRouter } from 'next/dist/client/router';
import { useEffect } from 'react';
import Error from '../components/error';
import Play from '../components/play/play';
import RoundEnd from '../components/round/round.end';
import RoundIntermission from '../components/round/round.intermission';
import RoundResult from '../components/round/round.result';
import { STATUS } from '../redux/game';
import { useAppDispatch, useAppSelector } from '../redux/hooks';
import { getRandomStreetView } from '../redux/position';
import { PageContentContainer } from '../styles';

const Game: NextPage = () => {
  const dispatch = useAppDispatch();
  const router = useRouter();

  const initialized = useAppSelector(({ game }) => game.initialized);
  const status = useAppSelector(({ game }) => game.status);
  const positionError = useAppSelector(({ position }) => position.error);

  useEffect(() => {
    if (!initialized) {
      router.push('/');
    }
  }, [initialized]);

  async function handleRetry() {
    await dispatch(getRandomStreetView({ radius: 1000 }));
  }

  function render() {
    switch (status) {
      case STATUS.INTERMISSION:
        return <RoundIntermission />;
      case STATUS.ROUND_STARTED:
        return <Play />;
      case STATUS.ROUND_ENDED:
        return <RoundEnd />;
      case STATUS.FINISHED:
        return <RoundResult />;

      default:
        return <div />;
    }
  }

  // If there is a header included, the page content container should
  // take the remaining height. The header is 48p by default
  /* 
   <GameHeader />
      <PageContentContainer height="calc(100% - 48px)">
        {render()}
      </PageContentContainer>
    </>
  
  */
  return (
    <>
      <PageContentContainer height="100%">
        {positionError ? (
          <Error
            callback={handleRetry}
            title="Error getting Street View data"
            info="This is likely because the map you chose is a little
                too small or has little Street View coverage."
            reason={positionError.message}
          />
        ) : (
          render()
        )}
      </PageContentContainer>
    </>
  );
};

export default Game;
