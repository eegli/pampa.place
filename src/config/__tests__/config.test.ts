import {config as gameConfig} from '../game';
import {config as googleConfig} from '../google';

describe('Game config', () => {
  jest.unmock('../game');

  it('includes unlimited time mode', () => {
    expect(gameConfig.timeLimits).toContain(-1);
  });
  it('defaults options are included by the config', () => {
    expect(gameConfig.rounds).toContain(gameConfig.roundsDefault);
    expect(gameConfig.timeLimits).toContain(gameConfig.timeLimitsDefault);
  });
  it('defines a marker color for each player', () => {
    expect(googleConfig.marker.colors.length).toBeGreaterThanOrEqual(
      gameConfig.maxPlayers
    );
  });
});
