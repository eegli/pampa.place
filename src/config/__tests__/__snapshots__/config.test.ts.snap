// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game config includes unlimited time mode 1`] = `
Object {
  "maxPlayers": 4,
  "rounds": Array [
    1,
    2,
    3,
  ],
  "roundsDefault": 2,
  "timeLimits": Array [
    30,
    60,
    120,
    -1,
  ],
  "timeLimitsDefault": 30,
}
`;
